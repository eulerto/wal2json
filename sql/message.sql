\set VERBOSITY terse

-- predictability
SET synchronous_commit = on;

CREATE TABLE message_table (a smallserial, b smallint);

SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');

-- One message
SELECT 'emit' FROM pg_logical_emit_message(true, 'foo', 'bar');
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);

-- Two messages
BEGIN;
SELECT 'emit' FROM pg_logical_emit_message(true, 'foo', 'bar');
SELECT 'emit' FROM pg_logical_emit_message(true, 'foo', 'bar');
COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);

-- Message sandwiched by changes
BEGIN;
INSERT INTO message_table (b) VALUES (1);
SELECT 'emit' FROM pg_logical_emit_message(true, 'foo', 'bar');
INSERT INTO message_table (b) VALUES (2);
COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);

-- Changes sandwiched by messages
BEGIN;
SELECT 'emit' FROM pg_logical_emit_message(true, 'before', 'before');
INSERT INTO message_table (b) VALUES (3);
SELECT 'emit' FROM pg_logical_emit_message(true, 'after', 'after');
COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);

-- Changes with newlines and quotes
SELECT 'emit' FROM pg_logical_emit_message(true, 'funky prefix', 'well " " " "" "" '' '' 
'' '' ] [ { } foo');
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);

-- Pretty
BEGIN;
SELECT 'emit' FROM pg_logical_emit_message(true, 'pretty prefix', 'pretty1');
SELECT 'emit' FROM pg_logical_emit_message(true, 'pretty prefix', 'pretty2');
COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'pretty-print', '1');

-- Small message renders correct size
SELECT 'emit' FROM pg_logical_emit_message(true, 'a pretty prefix that is long', 'a');
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'pretty-print', '1');

-- Non-transactional messages
BEGIN;
INSERT INTO message_table (b) VALUES (1);
SELECT 'emit' FROM pg_logical_emit_message(false, 'non-transactional prefix', 'non-transactional message');
INSERT INTO message_table (b) VALUES (2);
COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);

-- Standalone non-transactional messages
SELECT 'emit' FROM pg_logical_emit_message(false, 'non-transactional prefix', 'standalone non-transactional message');
SELECT pg_sleep(0.5); -- need a slight wait for the message to make it to decoding
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);

-- Bytea
SELECT 'emit' FROM pg_logical_emit_message(true, 'bytea prefix', decode('DEADBEEF', 'hex'));
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);

SELECT 'stop' FROM pg_drop_replication_slot('regression_slot');
