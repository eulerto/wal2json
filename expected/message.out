\set VERBOSITY terse
-- predictability
SET synchronous_commit = on;
CREATE TABLE message_table (a smallserial, b smallint);
SELECT 'init' FROM pg_create_logical_replication_slot('regression_slot', 'wal2json');
 ?column? 
----------
 init
(1 row)

-- One message
SELECT 'emit' FROM pg_logical_emit_message(true, 'foo', 'bar');
 ?column? 
----------
 emit
(1 row)

SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
                                        data                                         
-------------------------------------------------------------------------------------
 {"change":[{"kind":"message","transactional":true,"prefix":"foo","content":"bar"}]}
(1 row)

-- Two messages
BEGIN;
SELECT 'emit' FROM pg_logical_emit_message(true, 'foo', 'bar');
 ?column? 
----------
 emit
(1 row)

SELECT 'emit' FROM pg_logical_emit_message(true, 'foo', 'bar');
 ?column? 
----------
 emit
(1 row)

COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
                                                                            data                                                                            
------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"change":[{"kind":"message","transactional":true,"prefix":"foo","content":"bar"},{"kind":"message","transactional":true,"prefix":"foo","content":"bar"}]}
(1 row)

-- Message sandwiched by changes
BEGIN;
INSERT INTO message_table (b) VALUES (1);
SELECT 'emit' FROM pg_logical_emit_message(true, 'foo', 'bar');
 ?column? 
----------
 emit
(1 row)

INSERT INTO message_table (b) VALUES (2);
COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
                                                                                                                                                                               data                                                                                                                                                                                
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"change":[{"kind":"insert","schema":"public","table":"message_table","columnnames":["a","b"],"columntypes":["int2","int2"],"columnvalues":[1,1]},{"kind":"message","transactional":true,"prefix":"foo","content":"bar"},{"kind":"insert","schema":"public","table":"message_table","columnnames":["a","b"],"columntypes":["int2","int2"],"columnvalues":[2,2]}]}
(1 row)

-- Changes sandwiched by messages
BEGIN;
SELECT 'emit' FROM pg_logical_emit_message(true, 'before', 'before');
 ?column? 
----------
 emit
(1 row)

INSERT INTO message_table (b) VALUES (3);
SELECT 'emit' FROM pg_logical_emit_message(true, 'after', 'after');
 ?column? 
----------
 emit
(1 row)

COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
                                                                                                                                                    data                                                                                                                                                     
-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"change":[{"kind":"message","transactional":true,"prefix":"before","content":"before"},{"kind":"insert","schema":"public","table":"message_table","columnnames":["a","b"],"columntypes":["int2","int2"],"columnvalues":[3,3]},{"kind":"message","transactional":true,"prefix":"after","content":"after"}]}
(1 row)

-- Changes with newlines and quotes
SELECT 'emit' FROM pg_logical_emit_message(true, 'funky prefix', 'well " " " "" "" '' '' 
'' '' ] [ { } foo');
 ?column? 
----------
 emit
(1 row)

SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
                                                                  data                                                                  
----------------------------------------------------------------------------------------------------------------------------------------
 {"change":[{"kind":"message","transactional":true,"prefix":"funky prefix","content":"well \" \" \" \"\" \"\" ' ' \n' ' ] [ { } foo"}]}
(1 row)

-- Pretty
BEGIN;
SELECT 'emit' FROM pg_logical_emit_message(true, 'pretty prefix', 'pretty1');
 ?column? 
----------
 emit
(1 row)

SELECT 'emit' FROM pg_logical_emit_message(true, 'pretty prefix', 'pretty2');
 ?column? 
----------
 emit
(1 row)

COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'pretty-print', '1');
                        data                        
----------------------------------------------------
 {                                                 +
         "change": [                               +
                 {                                 +
                         "kind": "message",        +
                         "transactional": true,    +
                         "prefix": "pretty prefix",+
                         "content": "pretty1"      +
                 }                                 +
                 ,{                                +
                         "kind": "message",        +
                         "transactional": true,    +
                         "prefix": "pretty prefix",+
                         "content": "pretty2"      +
                 }                                 +
         ]                                         +
 }
(1 row)

-- Small message renders correct size
SELECT 'emit' FROM pg_logical_emit_message(true, 'a pretty prefix that is long', 'a');
 ?column? 
----------
 emit
(1 row)

SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL, 'pretty-print', '1');
                               data                                
-------------------------------------------------------------------
 {                                                                +
         "change": [                                              +
                 {                                                +
                         "kind": "message",                       +
                         "transactional": true,                   +
                         "prefix": "a pretty prefix that is long",+
                         "content": "a"                           +
                 }                                                +
         ]                                                        +
 }
(1 row)

-- Non-transactional messages
BEGIN;
INSERT INTO message_table (b) VALUES (1);
SELECT 'emit' FROM pg_logical_emit_message(false, 'non-transactional prefix', 'non-transactional message');
 ?column? 
----------
 emit
(1 row)

INSERT INTO message_table (b) VALUES (2);
COMMIT;
SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
                                                                                                                                            data                                                                                                                                            
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 {"change":[{"kind":"message","transactional":false,"prefix":"non-transactional prefix","content":"non-transactional message"}]}
 {"change":[{"kind":"insert","schema":"public","table":"message_table","columnnames":["a","b"],"columntypes":["int2","int2"],"columnvalues":[4,1]},{"kind":"insert","schema":"public","table":"message_table","columnnames":["a","b"],"columntypes":["int2","int2"],"columnvalues":[5,2]}]}
(2 rows)

-- Standalone non-transactional messages
SELECT 'emit' FROM pg_logical_emit_message(false, 'non-transactional prefix', 'standalone non-transactional message');
 ?column? 
----------
 emit
(1 row)

SELECT pg_sleep(0.5); -- need a slight wait for the message to make it to decoding
 pg_sleep 
----------
 
(1 row)

SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
                                                                    data                                                                    
--------------------------------------------------------------------------------------------------------------------------------------------
 {"change":[{"kind":"message","transactional":false,"prefix":"non-transactional prefix","content":"standalone non-transactional message"}]}
(1 row)

-- Bytea
SELECT 'emit' FROM pg_logical_emit_message(true, 'bytea prefix', decode('DEADBEEF', 'hex'));
 ?column? 
----------
 emit
(1 row)

SELECT data FROM pg_logical_slot_get_changes('regression_slot', NULL, NULL);
                                           data                                            
-------------------------------------------------------------------------------------------
 {"change":[{"kind":"message","transactional":true,"prefix":"bytea prefix","content":"Þ­"}]}
(1 row)

SELECT 'stop' FROM pg_drop_replication_slot('regression_slot');
 ?column? 
----------
 stop
(1 row)

